telosnetwork/telos-wallet
# Refactoring EVM Authentication (WalletConnect / Metamask / SafePal)

# Fixes #256 #422 #450

Note: I created a [new issue](https://github.com/telosnetwork/telos-wallet/issues/476) to finish some details left out of this PR.

## Description
This PR includes a complete refactoring of the authentication methods we already support with plans to add even more in the near future.

For this, a very similar strategy to the one we are already executing on the Native side was followed, but with a class structure more adapted to the EVM reality.

The following diagram shows the new structure under the name of EVMAuthenticator (at the right) and how it fits into the existing structure.

![evm-authenticators drawio](https://github.com/telosnetwork/telos-wallet/assets/4420760/9e736523-d6f5-4300-a2cd-73388fdad755)

## Screenshots
<table>
<tr><td><img src="https://github.com/telosnetwork/telos-wallet/assets/4420760/bc0d89cd-34d0-4fd0-a879-693cdb9e5dc8"></td>
<td><img src="https://github.com/telosnetwork/telos-wallet/assets/4420760/2bcdbd43-d411-49d2-9bf4-272363a6eb43"></td>
</tr>
</table>

## Test scenarios
The current solution (which does not have to be the definitive one) works in the following way. If the Metamask wallet is available, it is possible to detect and interact with it through the provider that the browser extension injects into the website under the name `window.ethereum`

However, if the SafePal wallet is active, regardless of whether Metamask is present or not, the injected object will be overwritten by the SafePal provider and therefore we will lose reference to the Metamask wallet (even though it is open and available).

### Login / Autologin && Logout
- Metamask
  - you need to have Metamask installed and available in your browser
  - if you have Safepal installed, you need to disable it for this test
- go to [this link](https://deploy-preview-465--wallet-staging.netlify.app/) and press "Connect your wallet" Button.
  - you should see the "Metamask" button, "Install SafePal" Button, and "WalletConnect" button
- press the Metamask button and follow the steps to log in
  - you should be successfully logged in and redirected to the balances page.
  - verify the account address is the same one you have in Metamask
  - all your balances, transaction history, and inventory items should be there with no changes
- press F5 standing on any page. You should be auto-logged with no changes.
- log out and press F5 to verify you are not auto-logged again
- now try  "Connect your wallet" and then "WalletConnect" button
  - only Metamask connector should appear.
  - Press it and follow the steps (if any) to log in. I you are already connected you will pass throw.
    - you should have exactly the same experience as connected directly with the Metamask wallet

- SafePal
  - you only need Safepal installed and enabled on your browser
- go to [this link](https://deploy-preview-465--wallet-staging.netlify.app/) and press "Connect your wallet" Button.
  - you should see the "Install Metamask" button, "SafePal" Button, and "WalletConnect" button
- press the SafePall button and follow the steps to log in
  - you should be successfully logged in and redirected to the balances page.
  - verify the account address is the same one you have in SafePal Wallet
  - all your balances, transaction history, and inventory items should be there with no changes
- press F5 standing on any page. You should be auto-logged with no changes.
- log out and press F5 to verify you are not auto-logged again
- now try  "Connect your wallet" and then "WalletConnect" button
  - both Metamask and SafePal connectors should appear (see screenshot)
  - Press the SafePal connector button and follow the steps (if any) to log in.
    - you should have exactly the same experience as connected directly with the SafePal wallet





## Checklist:
